core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -1320,6 +1331,214 @@ protected SqlNode performUnconditionalRewrites(
+          call.setOperand(i, newOperands.get(i));
+        }
+      }
+    } else if (current instanceof SqlNodeList) {
+      SqlNodeList nodeList = (SqlNodeList) current;
+      List<SqlNode> newNodes = new ArrayList<>();
+      for (int i = 0; i < nodeList.size(); i++) {
+        newNodes.add(
+            performHiveUdtfRewriteInternal(
+              nodeList.get(i), select2TableFunctionInfos));
+      }
+
+      for (int i = 0; i < newNodes.size(); i++) {
+        if (newNodes.get(i) != null) {
+          nodeList.set(i, newNodes.get(i));
+        }
+      }
+    }
+    return current;
+  }
+
+  /**
+   * Rewrite the "select a.id table_func(a.id) as (f0,f1)" to
+   * "select a.id,_table_function_0.f0, _table_function_0.f1"
+   * @param select SqlSelect Node
+   * @param selectItems select items to rewrite
+   * @param select2TableFunctionInfos Mapping of SqlSelect and it's TableFunction
+   * @return new SelectItems after rewrite
+   */
+  private SqlNodeList performRewriteForSelectItem(SqlSelect select, SqlNodeList selectItems,
+                             Map<SqlSelect, TableFunctionInfo> select2TableFunctionInfos) {
+    // step1. find the table function in the select items.
+    for (int i = 0; i < selectItems.size(); i++) {
+      SqlNode selectItem = selectItems.get(i);
+      if (selectItem.getKind() == SqlKind.AS) {
+        SqlNode udtfNode = ((SqlBasicCall) selectItem).getOperands()[0];
+        SqlNode aliasNode = ((SqlBasicCall) selectItem).getOperands()[1];
+
+        // test if this is a "table_func() as (f0,f1)" select item.
+        if (udtfNode instanceof SqlBasicCall
+            && ((SqlBasicCall) udtfNode).getOperator() instanceof SqlFunction
+            && aliasNode instanceof SqlNodeList) {
+
+          SqlFunction function = (SqlFunction)
+              ((SqlBasicCall) udtfNode).getOperator();
+          List<SqlOperator> overloads  = new ArrayList<>();
+          opTab.lookupOperatorOverloads(function.getNameAsId(),
+              SqlFunctionCategory.USER_DEFINED_TABLE_FUNCTION, SqlSyntax.FUNCTION, overloads);
+
+          if (overloads.size() == 0) {
+            throw newValidationError(udtfNode,
+                RESOURCE.exceptTableFunction(function.getName()));
+          }
+          // this is a table function
+          if (overloads.size() == 1 && overloads.get(0)
+              instanceof SqlUserDefinedTableFunction) {
+            //Only one table function allowed in select
+            if (select2TableFunctionInfos.containsKey(select)) {
+              throw newValidationError(udtfNode, RESOURCE.onlyOneTableFunctionAllowedInSelect());
+            }
+            TableFunctionInfo tableFunctionInfo = new TableFunctionInfo();
+            tableFunctionInfo.node = (SqlBasicCall) udtfNode;
+            tableFunctionInfo.selectIndex = i;
+            tableFunctionInfo.fieldNames = (SqlNodeList) aliasNode;
+            tableFunctionInfo.tableName = "_table_function_" + nextTableFunctionNameId++;
```

Comment:
make `"_table_function_"` a constant string may be better like 
`public static final String TABLE_FUNCTION_PREFIX = "_table_function_"`
The global variables `nextTableFunctionNameId` once use once in this function is a little odd

Reply:
I agree with you than make "_table_function_" as constant. But `nextTableFunctionNameId ` should be a global variable  in `SqlValidatorImpl`, as we may reuse the `SqlValidatorImpl` to validator different `SqlNode`. It is better to differentiate the table function in different `SqlNode`.